src/mainboard/starfive/visionfive2/bt0/src/main.rs

```diff
@@ -225,140 +277,85 @@ fn main() {
-                print!(".");
-            }
-        }
-        println!(" done.");
-        if DEBUG {
-            println!("Start:");
-            dump_block(dram, 0x100, 0x20);
-            println!("Presumably JH7110 recovery:");
-            dump_block(dram + 0x0002_0000, 0x100, 0x20);
-            println!("DTB:");
-            dump_block(dram + 0x0010_0000, 0x100, 0x20);
-            println!("Something:");
-            dump_block(dram + 0x0020_0000, 0x100, 0x20);
-        }
+    // Find and copy the main stage
+    let (base, size) = if LOAD_FROM_FLASH {
+        (QSPI_XIP_BASE, FLASH_SIZE)
     } else {
-        let base = 0x0800_0000 + 32 * 1024;
-        let dram = DRAM_BASE;
-        for b in (0..0x100).step_by(4) {
-            write32(dram + b, read32(base + b));
-            if b % 0x4_0000 == 0 {
-                print!(".");
-            }
+        (SRAM0_BASE, SRAM0_SIZE) // occupied space
+    };
+    let slice = unsafe { core::slice::from_raw_parts(transmute(base), size) };
+    let (main_offset, main_size) = get_main_offset_and_size(slice);
+    let main_addr = base + main_offset;
+
+    let load_addr = DRAM_BASE;
+
+    let main_size_k = main_size / 1024;
+    println!("Copy {main_size_k}k main stage from {main_addr:08x} to {load_addr:08x}... ⏳");
+    for b in (0..main_size).step_by(4) {
+        write32(load_addr + b, read32(main_addr + b));
+        if b % 0x4_0000 == 0 {
+            print!(".");
         }
-        println!("Payload:");
-        dump_block(dram, 0x20, 0x20);
+    }
+    println!(" done.");
+    if DEBUG {
+        dump_block(load_addr, 0x20, 0x20);
     }
 
-    println!("lzss compressed Linux");
-    dump_block(QSPI_XIP_BASE + 0x0040_0000, 0x100, 0x20);
-
-    println!("release non-boot harts =====\n");
-    {
+    // .....
+    if false {
+        println!("release non-boot harts =====\n");
         let clint = pac::clint_reg();
-        clint.msip_1().write(|w| w.control().set_bit());
+        clint.msip_0().write(|w| w.control().set_bit());
         clint.msip_2().write(|w| w.control().set_bit());
         clint.msip_3().write(|w| w.control().set_bit());
         clint.msip_4().write(|w| w.control().set_bit());
     }
 
-    println!("Jump to payload... with dtb {dtb:#x}");
-    exec_payload(dtb);
-    println!("Exit from payload, resetting...");
+    // GO!
+    println!("Jump to main stage @{load_addr:08x}");
+    exec_payload(load_addr);
+    println!("Exit from main stage, resetting...");
     unsafe {
         sleep(0x0100_0000);
         reset();
         riscv::asm::wfi()
     };
 }
 
-fn exec_payload(dtb: usize) {
-    let load_addr = if LOAD_FROM_FLASH {
-        // U-Boot proper expects to be loaded here
-        // see SYS_TEXT_BASE in U-Boot config
-        DRAM_BASE + 0x0020_0000
-    } else {
-        DRAM_BASE
-    };
-    // println!("Payload @{load_addr:08x}");
-
-    let hart_id = mhartid::read();
-    // U-Boot proper
+fn exec_payload(addr: usize) {
     unsafe {
```

Comment:
Similar to above, maybe add a `SAFETY` comment about `transmute`ing executable memory into an `EntryPoint`?

Since we're not doing a signed/secure boot setup, we're assuming the memory is valid, but I think that should at least be mentioned.